### ISSUE TITLE
incorrect utf-8 being serialized for 1F4A1 and perhaps others

### ISSUE THREAD

**Issue by devnut** [rails/rails#24842]
### Steps to reproduce
hash = {'str' => '💡'}
hash['str'].bytes do |b|
  puts "#{b.to_s(16)}\n"
end
f0
9f
92
a1

...This is correct so far, but then it does something unexpected:

2.1.1 :134 > hash['str'].to_json
 => "\"\\uf4a1\"" 

This does not appear to be correct.  It is code 1F4A1, perhaps that is what is intended?

### Expected behavior

Should return the lightbulb utf-8 character.

### Actual behavior

hash = {'str' => '💡'}
broken_hash = JSON.parse(hash.to_json)
broken_hash['str'].bytes do |b|
  puts "#{b.to_s(16)}\n"
end
ef
92
a1

### System configuration
**Rails version**: 3.2.22.2

**Ruby version**:2.1.1 (tested on other versions, this is just what I'm running it on at the moment)

**Comment by vipulnsward** [comment]
This works as expected on master. Going to give it a close. 
```ruby
irb(main):009:0> broken_hash = JSON.parse(hash.to_json)
=> {"str"=>"💡"}
irb(main):010:0> broken_hash['str'].bytes do |b|
irb(main):011:1* puts "#{b.to_s(16)}\n"
irb(main):012:1> end
f0
9f
92
a1
=> "💡"
```

@devnut Thanks for reporting.

**Comment by devnut** [comment]
Is there any chance we can have whatever fixed this back-ported into 3.2-stable?
